### DISCUSSION TITLE
Breadcrumbs

### DISCUSSION THREAD

**Discussion by juledwar** [zauberzeug/nicegui#1181] [Q&A] [answered]
### Question

I'm pretty new to this project but loving it so far!

I noticed that the Quasar breadcrumb component is not implemented (https://quasar.dev/vue-components/breadcrumbs/) so I tried doing something funky with labels and expansions and didn't get very var (I should note that I'm not really a front end dev so I may be doing it all wrong anyway).

Is it going to be possible to do a breadcrumb navigation something like this?
```
Thing1 v / Thing2 v / Thing3 
```
Where each breadcrumb is a clickable bit of text and an expansion.

So far I did something like this:
```py
       with ui.expansion() as expansion:
            expansion.props('dense dense-toggle')
            with expansion.add_slot('header'):
                thing1_label=ui.label('Thing1').classes('underline cursor-pointer self-center')
                thing1_label.on('click', do_something)
            inside_label=ui.label('foo').classes('underline cursor-pointer self-center')
            inside_label.on('click', do_something)

etc...
```

I'm trying to make the "Thing" texts and expansion items clickable so I can call a function off a click event, but the click event doesn't happen. Similarly for any labels inside the expansion.

Thanks for any pointers!

**Comment by falkoschindler** [comment, marked as answer]
Hi @juledwar,

Happy to hear that you like the project. 😀

I'm not sure if I understand your problem correctly. Even though there is no dedicated breadcrumbs element in NiceGUI, you could create one like this:
```py
with ui.element('q-breadcrumbs'):
    ui.element('q-breadcrumbs-el').props('label=foo icon=home')
    ui.element('q-breadcrumbs-el').props('label=bar')
    ui.element('q-breadcrumbs-el').props('label=buz')
```
You could style it using `.style()` and `.classes()` and add click handlers with `.on('click', ...)`.

But somehow you want expansion elements. So your approach might be better suited.

You write "the click event doesn't happen". When I try your code with
```py
def do_something():
    print('something')
```
I see "something". Are you sure the click event doesn't work?

**Reply by juledwar** [reply]
Hi there!

Oh I didn't realize I could use `ui.element` like that, I might give it a go, thanks.

However, as you note, I wanted to make each crumb an expansion element with clickable links to navigate to alternative paths and as a newbie to this I'm struggling to understand the best way.

As for the click event, I'm absolutely sure it's not working. I changed the callback to a `ui.notify` just to make sure and nothing appears. All that happens is that the expansion opens and closes for `thing_label` clicks, and nothing at all for `inside_label`.  Weird.

**Reply by falkoschindler** [reply]
Strange. When clicking on `thing1_label`, the event is called. Maybe you're clicking somewhere next to the `thing1_label` (but still inside the expansion). So maybe you're looking for something like this:

```py
with ui.expansion().on('click', do_something) as expansion:
    with expansion.add_slot('header'):
        ui.label('Thing1').classes('underline cursor-pointer self-center')
    ui.label('foo').classes('underline cursor-pointer self-center').on('click', do_something)
```

**Reply by juledwar** [reply]
I copy-pasted this exact code to a file called `test.py`, changed the `do_something` for a `ui.notify("foo")`, added a `ui.run()` at the bottom and ran it with `python3 test.py`. I get no notifications when clicking either the expansion or the inner label.

:(

**Reply by falkoschindler** [reply]
Just to be sure: Did you write `ui.expansion().on('click', ui.notify("foo"))` or `ui.expansion().on('click', lambda: ui.notify("foo"))`? Only the latter will work.

**Reply by juledwar** [reply]
Oh man, that's embarrassing... I seem to get what I need now, thanks! Now, to work out how to style it ....

Thank you!

**Comment by juledwar** [comment]
It turns out that an expansion is not really what I need as I need to dynamically generate the content. The only component I can see that allows this is the `select` one, is that right? The Quasar https://quasar.dev/vue-components/button-dropdown/ component would potentially be perfect but I don't think that's implemented here.

Is there going to be any way to do what I described in the main post above, with dynamically generated breadcrumb components? I'm starting to doubt it :disappointed:

**Reply by falkoschindler** [reply]
Here is an example for QBtnDropdown in NiceGUI:
```py
with ui.element('q-btn-dropdown').props('label="Dropdown Button" color=primary') as dropdown:
    with ui.element('q-list'):
        with ui.element('q-item').props('clickable').on('click', lambda: ui.notify('Clicked Photos')):
            with ui.element('q-item-section'):
                ui.label('Photos')
        with ui.element('q-item').props('clickable').on('click', lambda: ui.notify('Clicked Videos')):
            with ui.element('q-item-section'):
                ui.label('Videos')
        with ui.element('q-item').props('clickable').on('click', lambda: ui.notify('Clicked Articles')):
            with ui.element('q-item-section'):
                ui.label('Articles')
```
Of course you can add and remove items dynamically: https://nicegui.io/documentation#clear_containers

**Reply by juledwar** [reply]
Falko, thanks VERY much for this example, this has really helped my understanding of the project generally as well as solved my immediate problem. I've now built exactly what I wanted with minimal fuss. Fantastic!

**Comment by belamu** [comment]
I would also like to use breadcrumbs for a hierachy of things (locations) and a click on one crumb should go there. From my impression this is the basic usage of breadcrumbs but I cannot figure out how to add the links. My try:

```python
@ui.page("/location")
def location_page(levels: Annotated[list[myValue] | None, fastapi.Query()] = None):
    """Display location editor.
    Args:
        levels: the location specified by the hierarchy of levels
    """
    levels = [] if levels is None else levels
    with ui.element("q-breadcrumbs"):
        for nr, level in enumerate(levels):
            ui.element("q-breadcrumbs-el").props(f"label={level}").on(
                "click", lambda: ui.navigate.to(lambda: locations_page(levels[: nr + 1])))
    with ui.element("div"):
        my_locations_editor(levels)
```
but this fails with `KeyError: <function locations_page.<locals>.<lambda>.<locals>.<lambda> at 0x7fffe44b2770>`.
I guess the lambda wrapping the locations_page call is not a decorated page function and therefore fails.

In #2394 the lambda function was a solution to arguments in callback-functions but here it fails.

Probably I could create the url with the levels as `&levels=...`-arguments but this sounds much less clear.

Any help would be appreciated!

**Reply by belamu** [reply]
My workaround with url-creation now works but I don't like it really because the variable names of the page function and its argument reappear as strings and I have an additional import:

``` python
...).on(
   "click",
   lambda l=levels[: nr + 1]: ui.navigate.to(
       f"/location?{urllib.parse.urlencode([('levels', level) for level in l])}"
    ),
```
